Feat/serde enum representation

AlisCode opened the pull request:
This brings support for all kinds of [serde enum reprensentations](https://serde.rs/enum-representations.html).

Unit tests included, on top of which we tried generating Typescript bindings using openapi-generator (typescript-axios generator, not because of `Axios` but because the generated types are simple to understand). They do indeed look good, so we're confident that this works mostly like it should.

Maybe some missing docs? Could be done in a separate PR? If you feel like it, please provide examples of what needs an update.

Fixes #320 
Fixes #219

juhaku reviewed (changes requested):
This is great. Thank you for this patch. :1st_place_medal: 

Few things to clarify and we can get this thing going on. :slightly_smiling_face: 

Doc updates can be a seprate PR and also could be done here too. I guess it does not really matter that much :smile: 

Anyhow the thing that need some changes in docs reside in `utoipa-gen/src/lib.rs` file. And in line 155 there is a section about partial serde attribute support. 

It seems like this would work with `IntoParams` as well. The thing that is used to create parameters for the `utoipa::path` since they are actually using the same code. That part starts with 1482 line.  Probably would be useful to write a test at least for one of the new serde things for `IntoParams`  Tests regarding `IntoParams` can be found from `tests/path_derive.rs`

You can generage the docs with command:
```bash
cargo +nightly doc -Z unstable-options --workspace --no-deps --features utoipa/actix_extras,utoipa/openapi_extensions,utoipa/yaml,uuid,actix-web,axum,rocket --config 'build.rustdocflags = ["--cfg", "doc_cfg"]'
``` 

And then you can use your favourite browser to browse the generated docs. Found in the target folder.

For what comes to the docs. It might be handy to add a few examples regarding these tag attributes. Maybe? :thinking:

AlisCode commented:
Hey, sorry it took so long. I hope you get the time to review this again so we can get some progress here :) 

> It seems like this would work with IntoParams as well. The thing that is used to create parameters for the utoipa::path since they are actually using the same code.

Could you clarify what you mean ? It seems to me like `IntoParams` only supports `struct`, as expressed in `IntoParams::get_struct_fields` 

```rust
match &self.data {
    Data::Struct(data_struct) => { // ... [omitted] },
    _ => abort("Only struct type is supported"),
}
```

Therefore I'm not sure what impact this PR has ?

juhaku commented:
> Hey, sorry it took so long. I hope you get the time to review this again so we can get some progress here :)
> 
> > It seems like this would work with IntoParams as well. The thing that is used to create parameters for the utoipa::path since they are actually using the same code.
> 
> Could you clarify what you mean ? It seems to me like `IntoParams` only supports `struct`, as expressed in `IntoParams::get_struct_fields`
> 
> ```rust
> match &self.data {
>     Data::Struct(data_struct) => { // ... [omitted] },
>     _ => abort("Only struct type is supported"),
> }
> ```
> 
> Therefore I'm not sure what impact this PR has ?

Now that you mentioned. Just forget me saying about the `IntoParams` support. :smile: 

If there ever comes a need support these enums in `IntoParams` it can be adressed in future.

juhaku commented:
> Hey, sorry it took so long. I hope you get the time to review this again so we can get some progress here :)

No worries. I have also been quite busy last few weeks with life. I would say that addressing the comments in the code and adding docs either with separate PR will complete the feature set. 

> Could you clarify what you mean ? It seems to me like IntoParams only supports struct, as expressed in IntoParams::get_struct_fields
> Therefore I'm not sure what impact this PR has ?

To be clear the mention about `IntoParams` can be ignored in my original comment. I already lined it out.
> It seems like this would work with IntoParams as well. The thing that is used to create parameters for the utoipa::path since they are actually using the same code. That part starts with 1482 line. Probably would be useful to write a test at least for one of the new serde things for IntoParams Tests regarding IntoParams can be found from tests/path_derive.

Honestly I don't remember how did I get to think that that there is something to be done for `IntoParams`. :smile: Actually this PR itself will add support for `IntoParams` in a way that if an enum is used as parameter type `ToSchema::schema` will be called for the enum to generate the schema for the parameter.  Thus only place where these enum variants should be supported is in `ToSchema` implementation as you here have had implemented. The impact this PR gives is it just enhances `ToSchema` enums which itself is correct. 

`IntoParams` does not need itself support for enums as you found out as well the struct type is supported only. If there ever will be a need to support enums in `IntoParams` that should actually be another PR then. 

Hope this already clears it out a bit.  Just as if there is something unclear. It would be awesome to get this to the lib. :slightly_smiling_face:

AlisCode commented:
Just rebased and pushed two commits : 
* One fixes the docs to mention the new supported features and correct some english wording that sounds better IMHO
* One makes use of `abort!` instead of `panic!`, great catch, I didn't know that  

I struggle to get what looks like two very valid comments about the parsing of features. I'm not suuuper familiar with OpenAPI itself. 

For now on my local version I just removed the declaration of `_title_features`, because said features are unused. Tests are passing. Unclear if that means the `Title` feature shouldn't be allowed at all ?

By the way, even on master I have tests that didn't pass : 

```
Couldn't compile the test.

failures:
    src/lib.rs - into_params (line 1538)
    src/lib.rs - path (line 1066)
    src/lib.rs - path (line 1085)
    src/lib.rs - path (line 884)
    src/lib.rs - path (line 905)
```

Is that expected ?

AlisCode commented:
Just pushed this. I think this is good for re-review now

AlisCode commented on utoipa-gen/src/component/schema.rs:
```diff
@@ -484,9 +487,9 @@ impl ToTokens for UnnamedStructSchema<'_> {
-   pub enum_name: Cow<'a, str>,
-   pub variants: &'a Punctuated<Variant, Comma>,
-   pub attributes: &'a [Attribute],
+    pub enum_name: Cow<'a, str>,
+    pub variants: &'a Punctuated<Variant, Comma>,
+    pub attributes: &'a [Attribute],
```
This was triggered by my neovim having format-on-save. `cargo fmt` doesn't revert this back so I'm assuming it's correct? 

I can probably try to revert it if you want, in any case, I don't think it's very relevant

juhaku replied:
Okay, it might be that the format there is incorrect in first place.

juhaku replied:
Yeah it seems like so. It's fine :slightly_smiling_face:

juhaku commented:
> By the way, even on master I have tests that didn't pass :

This might be because of some of the tests needs cargo features not enabled by default. if you run the tests with `./scripts/test.sh utoipa-gen` and `./scripts/test.sh utoipa` They should pass.

juhaku commented:
>For now on my local version I just removed the declaration of _title_features, because said features are unused. Tests are passing. Unclear if that means the Title feature shouldn't be allowed at all ?

The title feature is bit special case. In OpenAPI you can specify title for the each schema object. The `split_for_title` is to extract `title` attribute from derive attributes of enum variant. In the illustration below the `title` should be placed into enum variant `Bar` object and not into `Foo` object. 
```rust
#[derive(ToSchema)]
struct Foo {}

#[derive(ToSchema)]
enum FooBar {
  #[schema(title  = "Bar")]
  Bar(Foo),
}
```
When the enum variant is created with `ObjectVariant` we add the possible title to the enum object variant itself. If we did not call `split_for_title` method the `Title` feature containing `title` attribute will be passed to the `UnnamedStructSchema` within the `unnamed_struct_features` which in turn will result incorrect OpenAPI spec what cannot be parsed correctly by the client generation libraries. 
```rust
ObjectVariant {
  title: title_features.first().map(ToTokens::to_token_stream),
  item: UnnamedStructSchema {
      struct_name: Cow::Borrowed(self.enum_name),
      attributes: &variant.attrs,
      features: Some(unnamed_struct_features),
      fields: &unnamed_fields.unnamed,
  },
}
```

juhaku reviewed (approved):
This looks pretty good :+1: :trophy:  Let's make the ball rolling.

AlisCode commented:
Thanks for the merge and the good work on the lib ! :)
